exotations/solvers/exotica_control_rrt_solver/include/exotica_control_rrt_solver/control_rrt_solver.h

```diff
@@ -0,0 +1,135 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#ifndef EXOTICA_CONTROL_RRT_SOLVER_CONTROL_RRT_SOLVER_H_
+#define EXOTICA_CONTROL_RRT_SOLVER_CONTROL_RRT_SOLVER_H_
+
+// #include <exotica_core/feedback_motion_solver.h>
+#include <exotica_core/motion_solver.h>
+#include <exotica_core/problems/dynamic_time_indexed_shooting_problem.h>
+
+#include <exotica_control_rrt_solver/control_rrt_solver_initializer.h>
+
+// TODO: Remove unused includes
+#include <ompl/base/goals/GoalState.h>
+#include <ompl/base/spaces/SE2StateSpace.h>
+#include <ompl/control/SimpleSetup.h>
+#include <ompl/control/SpaceInformation.h>
+#include <ompl/control/planners/rrt/RRT.h>
+#include <ompl/control/spaces/RealVectorControlSpace.h>
+
+namespace ob = ompl::base;
+namespace oc = ompl::control;
+
+namespace exotica
+{
+class OMPLStatePropagator : public oc::StatePropagator
+{
+public:
+    OMPLStatePropagator(
+        oc::SpaceInformationPtr si,
+        DynamicsSolverPtr dynamics_solver_) : oc::StatePropagator(si), space_(si), dynamics_solver_(dynamics_solver_) {}
+    void propagate(
+        const ob::State *state,
+        const oc::Control *control,
+        const double duration,
+        ob::State *result) const override
+    {
+        double t = timeStep_;
+        space_->copyState(result, state);
+
+        while (t < duration + std::numeric_limits<double>::epsilon())
+        {
+            Integrate(result, control, timeStep_);
+            t += timeStep_;
+        }
+    }
+
+    void setIntegrationTimeStep(double timeStep)
+    {
+        timeStep_ = timeStep;
+    }
+
+    double getIntegrationTimeStep() const
+    {
+        return timeStep_;
+    }
+
+private:
+    double timeStep_ = 0.0;
+    oc::SpaceInformationPtr space_;
+    DynamicsSolverPtr dynamics_solver_;
+
+    void Integrate(ob::State *ob_x, const oc::Control *oc_u, double dt) const
+    {
+        const int NU = dynamics_solver_->get_num_controls();
+        const int NX = dynamics_solver_->get_num_positions() + dynamics_solver_->get_num_velocities();
+
+        double *x = ob_x->as<ob::RealVectorStateSpace::StateType>()->values;
+        double *u = oc_u->as<oc::RealVectorControlSpace::ControlType>()->values;
+
+        Eigen::VectorXd eig_x = Eigen::Map<Eigen::VectorXd>(x, NX);
+        Eigen::VectorXd eig_u = Eigen::Map<Eigen::VectorXd>(u, NU);
```

Comment:
Maybe add an assert (so will only work in debug) that the size of ob::State is the same as NX etc.)

Reply:
What do you mean only work in debug?

Reply:
If you use `assert`, it will only be executed when compiled in Debug. Assert statements are removed from release. So it's good to test something but without affecting performance in Release.

Reply:
added

Reply:
Actually, this doesn't work and I'm not sure it can. See
https://ompl.kavrakilab.org/classompl_1_1base_1_1RealVectorStateSpace_1_1StateType.html#details

I don't know how to cast from `RealVectorStateSpace::StateType` back to `RealVectorStateSpace` or if that's even possible..